linker/shared/src/main/scala/org/scalajs/linker/OutputMode.scala

```diff
@@ -0,0 +1,99 @@
+/*                     __                                               *\
+**     ________ ___   / /  ___      __ ____  Scala.js tools             **
+**    / __/ __// _ | / /  / _ | __ / // __/  (c) 2013-2015, LAMP/EPFL   **
+**  __\ \/ /__/ __ |/ /__/ __ |/_// /_\ \    http://scala-js.org/       **
+** /____/\___/_/ |_/____/_/ | |__/ /____/                               **
+**                          |/____/                                     **
+\*                                                                      */
+
+
+package org.scalajs.linker
+
+/** JavaScript output mode configuration.
+ *
+ *  The output mode specifies what features of modern versions of JavaScript
+ *  are used by the Scala.js linker.
+ *
+ *  - Options whose name is of the form `useX` *force* the linker to use the
+ *    corresponding features, guaranteeing that the specific semantics that
+ *    they provide will be used.
+ *  - Options whose name is of the form `allowX` *allow* the linker to use the
+ *    corresponding features if it supports them. Support for such options can
+ *    be dropped in any subsequent version of the linker, including patch
+ *    versions.
+ */
+final class OutputMode private (
+    /** Whether to use ECMAScript 2015 features, such as classes and arrow
+     *  functions.
+     */
+    val useECMAScript2015: Boolean
+) {
+  import OutputMode._
+
+  private def this() = {
+    this(
+        useECMAScript2015 = false
+    )
+  }
+
+  def withUseECMAScript2015(useECMAScript2015: Boolean): OutputMode =
+    copy(useECMAScript2015 = useECMAScript2015)
+
+  override def equals(that: Any): Boolean = that match {
+    case that: OutputMode =>
+      this.useECMAScript2015 == that.useECMAScript2015
+    case _ =>
+      false
+  }
+
+  override def hashCode(): Int = {
+    import scala.util.hashing.MurmurHash3._
+    var acc = HashSeed
+    acc = mixLast(acc, useECMAScript2015.##)
+    finalizeHash(acc, 1)
+  }
+
+  override def toString(): String = {
+    s"""OutputMode(
+       |  useECMAScript2015 = $useECMAScript2015
+       |)""".stripMargin
+  }
+
+  private def copy(
+      useECMAScript2015: Boolean = this.useECMAScript2015
+  ): OutputMode = {
+    new OutputMode(
+        useECMAScript2015 = useECMAScript2015
+    )
+  }
+}
+
+object OutputMode {
+  private val HashSeed =
+    scala.util.hashing.MurmurHash3.stringHash(classOf[OutputMode].getName)
+
+  /** Default configuration of the output mode.
+   *
+   *  - `useECMAScript2015`: false
+   */
+  val Defaults: OutputMode = new OutputMode()
+
+  @deprecated("Use Defaults instead.", "1.0.0")
```

Comment:
This should be deprecated in "1.0.0-M4" instead, no? Ideally we can remove this in 1.0.0

Reply:
If we want to remove in 1.0.0, there needs to be a replacement available in 0.6.x. We could kind of do it. For `Default`/`Defaults` it is easy. For the `withX` and `x` methods it's also possible, for example:
```scala
sealed abstract class OutputMode {
  def useECMAScript2015: Boolean = this match {
    case ECMAScript51Isolated => false
    case ECMAScript6           => true
  }

  def withUseECMAScript2015(useECMAScript2015: Boolean): OutputMode =
    if (useECMAScript2015) ECMAScript6
    else ECMAScript51Isolated
}
```
Should we do that in the 0.6.x branch?

Reply:
If you don't mind. I think it is worth going through that extra little bit to get the legacy out of 1.0.

Reply:
➡️ #3305